torch/_dynamo/repro/after_aot.py

```diff
@@ -647,3 +601,277 @@ def inductor_accuracy_fails(fx_g, args, check_str=None):
+    torch._inductor.config.generate_intermediate_hooks = True
+
+    if options.minify:
+        from functorch.compile import minifier
+
+        favored_device = 1 if torch.cuda.device_count() >= 2 else 0
+        env_variables = {"CUDA_VISIBLE_DEVICES": str(favored_device)}
+
+        module_fails: Any
+        if options.isolate:
+            module_fails = functools.partial(
+                isolate_fails,
+                "inductor_accuracy_fails" if options.accuracy else "inductor_fails",
+                env=env_variables,
+                compiler_name=compiler_name,
+                patch_code=options.patch_code,
+                save_dir=options.save_dir,
+            )
+        else:
+            module_fails = (
+                inductor_accuracy_fails if options.accuracy else inductor_fails
+            )
+
+        minifier(
+            mod,
+            args,
+            module_fails=module_fails,
+            dump_state=functools.partial(
+                dump_compiler_graph_state, compiler_name=compiler_name
+            ),
+        )
+
+    elif options.analyze:
+        # TODO: The logic for cloning inputs/models here is intentionally
+        # modeled off of run_fwd_maybe_bwd, but arguably it is better not to
+        # clone inputs (as you are doubling your effective GPU memory usage).
+        # It is certainly faster though!  It probably makes sense to let the
+        # user specify the offload strategy.
+        from torch._inductor.hooks import intermediate_hook
+
+        with tqdm(desc="Compiling"):
+            compiled = compile_fx_inner(mod, args)
+        total = counters["inductor"]["intermediate_hooks"]
+
+        known_names = set()
+
+        def save_hook(name, val):
+            known_names.add(name)
+            if not options.skip_saving_inductor_intermediates:
+                writer.write_tensor(os.path.join("inductor", name), val)
+            pbar.update(1)
+
+        writer = torch.utils._content_store.ContentStoreWriter(
+            options.save_dir, stable_hash=options.stable_hash
+        )
+        reader = torch.utils._content_store.ContentStoreReader(options.save_dir)
+
+        new_args = clone_inputs(args)
+        with intermediate_hook(save_hook), tqdm(
+            desc="Saving inductor intermediates", total=total
+        ) as pbar:
+            compiled(new_args)
+            assert not new_args
+
+        def compare_tuples(tuple1, tuple2):
+            diff_indices = [i for i in range(len(tuple1)) if tuple1[i] != tuple2[i]]
+            diff_values = [(tuple1[i], tuple2[i]) for i in diff_indices]
+
+            if not diff_values:
+                return None
+            else:
+                return " and ".join(f"{a} != {b}" for a, b in diff_values)
+
+        def check_hook(name, val):
+            meta = writer.compute_tensor_metadata(val)
+            meta2 = reader.read_tensor_metadata(os.path.join("inductor", name))
```

Comment:
Is OOM'ing due to reading the previous intermediate into memory while we're in the middle of execution a concern? We could probably make that less likely by just writing intermediates on both runs of the model, and then looping through the intermediates afterwards and reading them into memory + comparing them one at a time. That's definitely slower though, so I don't know if it's worth it.

Reply:
The current analysis memory usage is not optimal, but it is good enough for the particular case I was working on. My general attitude is that, if we discover we need the more memory efficient variants, we should implement them, but we always want to keep the faster more memory guzzling versions around, as they're much more pleasant when they work.